lib/components/viewers/route-viewer.js

```diff
@@ -51,55 +79,121 @@ class RouteViewer extends Component {
       transitOperators,
       viewedRoute
     } = this.props
-    const sortedRoutes = routes
-      ? Object.values(routes).sort(
-        coreUtils.route.makeRouteComparator(transitOperators)
+    const { agency, mode, search } = this.state
+    const routesArray = routes ? Object.values(routes) : []
+
+    const filteredAndSortedRoutes = routesArray
+      .filter(
+        (route) =>
+          // If the filter isn't defined, don't check.
+          (!agency || agency === route.agencyName) &&
+          (!mode || mode === route.mode) &&
+          // If user search is active, filter here
+          (!search ||
+            ((route.longName && route.longName.includes(search)) ||
+              (route.shortName && route.shortName.includes(search))))
+      )
+      .sort(coreUtils.route.makeRouteComparator(transitOperators))
+
+    // get agency and mode lists from route list
+    const agencies = Array.from(
+      new Set(routesArray.map((route) => route.agencyName || route.agency.name))
+    )
+      .filter((agency) => agency !== undefined)
+      .sort()
+
+    const modes = Array.from(
+      new Set(
+        routesArray
+          .filter((route) => route.mode !== undefined)
+          .map((route) => route.mode.toLowerCase())
       )
-      : []
+    )
+      // Correct capitalization
+      .map((mode) => mode[0].toUpperCase() + mode.substr(1).toLowerCase())
+      .sort()
+
     return (
       <div className='route-viewer'>
         {/* Header Block */}
         <div className='route-viewer-header'>
           {/* Back button */}
           {!hideBackButton && (
             <div className='back-button-container'>
-              <Button
-                bsSize='small'
-                onClick={this._backClicked}
-              ><Icon type='arrow-left' />Back</Button>
+              <Button bsSize='small' onClick={this._backClicked}>
+                <Icon type='arrow-left' />
+                Back
+              </Button>
             </div>
           )}
 
           {/* Header Text */}
           <div className='header-text'>
             {languageConfig.routeViewer || 'Route Viewer'}
           </div>
-          <div className=''>
-            {languageConfig.routeViewerDetails}
-          </div>
-          <div style={{ clear: 'both' }} />
+          <div className=''>{languageConfig.routeViewerDetails}</div>
+          <section className='search-and-filter'>
```

Comment:
For this section, I think the mobile view should be how things look on desktop too. I.e., two rows (top with selects, bottom with text search). Also, can we extend the agency filter to fill the empty space (since oftentimes, the agency names are quite long and get cut off)?
![image](https://user-images.githubusercontent.com/2370911/130126345-c8467ab6-a887-4427-b0f0-49f2618913e9.png)

Reply:
All possible and all done in f638c7f23bedaa87c1472472090145f129563cca